### PR TITLE
[2.x] Add integration tests using fixtures files.

### PR DESCRIPTION
**Pull request by SpacePossum**
There are good unit tests for fixers (and other parts of the code),
but there is no easy way to make integration tests.
  
This PR adds integration tests to the PHP-CS-Fixer code base.
It aims to makes it easy to tests (whole) PHP code (files) using multiple fixers.

Inspired by the fixtures in Twig.

Limitations:
* it doesn't test/do linting (by design)
* file based fixers won't work (like PSR0 and possibly PSR4 if that fixer is made)

A `.test` has the following structure:
```
--TEST--
My test title
--CONFIG--
level=none|psr0|psr1|psr2|symfony
fixers=name1,name2 (additonal fixers)
--fixers=name3,name4 (fixers not to use)
--INPUT--
some code
--EXPECT--
expected code after fixing
```

config items `fixers` and `--fixers` are optional
`--EXPECT--` is optional, if omitted the input is expected to stay the same.

Examples can be seen here:
* https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1375/files#diff-e283518231c001f9337e0093560cb730R1
* https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1375/files#diff-acdfc6b1dc6d83c7a9ae915af8acb5a1R1

### CODE REVIEW

File: Symfony/CS/Fixer.php [hunk 1]
```diff
@@ -82,13 +89,17 @@ public function __construct()
     public function registerBuiltInFixers()
     {
         foreach (Finder::create()->files()->in(__DIR__.'/Fixer') as $file) {
+            /* @var FinderSplFileInfo $file */
```
Comment (GrahamCampbell) [inline Symfony/CS/Fixer.php]: wrong

Comment (GrahamCampbell) [inline Symfony/CS/Fixer.php]: this needs to be ABOVE the foreach, and MUST be `/**` style, otherwise it's not even phpdoc

File: Symfony/CS/Fixer.php [hunk 2]
```diff
@@ -110,6 +124,7 @@ public function getFixers()
     public function registerBuiltInConfigs()
     {
         foreach (Finder::create()->files()->in(__DIR__.'/Config') as $file) {
+            /* @var FinderSplFileInfo $file */
```
Comment (GrahamCampbell) [inline Symfony/CS/Fixer.php]: same

File: Symfony/CS/Fixer.php (hunk 1 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer.php]: thanks, please see: /4e03f7b7229486cb689362d9205b1128521dc619

File: Symfony/CS/Fixer.php (hunk 2 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer.php]: same, thanks, please see: /4e03f7b7229486cb689362d9205b1128521dc619

**Comment by GrahamCampbell** [comment]
Surely all the phpdoc changes should be made on the 1.11 branch?

**Comment by SpacePossum** [comment]
I guess :+1: , lets see if this is a feature wanted before getting all the details in place :)

**Comment by GrahamCampbell** [comment]
Sure, but the phpdoc additions are mainly unrelated, which is why I'm suggesting taking them out of this PR. :)

File: Symfony/CS/Fixer.php [hunk 3]
```diff
@@ -311,6 +312,48 @@ public function fixFile(\SplFileInfo $file, array $fixers, $dryRun, $diff, FileC
...
     }
 
+    /**
+     * @param \SplFileInfo     $file
+     * @param Tokens           $tokens
+     * @param FixerInterface[] $fixers
+     *
+     * @return string[]
+     */
+    public function fixFileTokens(\SplFileInfo $file, Tokens $tokens, array $fixers)
```
Comment (stof) [inline Symfony/CS/Fixer.php]: why is this public ? It does not make sense to have it as part of the public API IMO

File: Symfony/CS/Tests/IntegrationTest.php [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+     * @var Fixer
+     */
+    private $fixer;
+
+    /**
+     * @dataProvider getTests
+     */
+    public function testIntegration($fileName, $testTitle, $fixers, $input, $expected = null)
+    {
+        $this->fixer = new Fixer(); // cannot be done in setUp(): https://github.com/sebastianbergmann/phpunit/issues/836
```
Comment (stof) [inline Symfony/CS/Tests/IntegrationTest.php]: the issue you link has nothing to do with setUp. It talks about setUpBeforeClass. So the comment is wrong

File: Symfony/CS/Fixer.php (hunk 3 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer.php]: A alternative would be to mock the Fixer in the test and up the visibility of the method there,
would that be more acceptable?

**Comment by stof** [comment]
Why making this a 2.0 feature ? New features can (and should) go in 1.x.

I also think you should separate the logic allowing to create integration tests from the actual tests using it in this library, to make the integration testing setup reusable by other packages providing extra fixers. See how Twig does it.

File: Symfony/CS/Fixer.php [hunk 5]
```diff
@@ -286,12 +287,12 @@ public function fixFile(\SplFileInfo $file, array $fixers, $dryRun, $diff, FileC
                 return;
             }
 
-            if (!$dryRun && false === @file_put_contents($file->getRealpath(), $new)) {
+            if (!$dryRun && false === @file_put_contents($file->getRealPath(), $new)) {
```
Comment (stof) [inline Symfony/CS/Fixer.php]: fixing the case of the method name is unrelated to adding integration tests. You should submit it as a separate PR

File: Symfony/CS/Fixer.php [hunk 6]
```diff
@@ -121,6 +136,9 @@ public function addConfig(ConfigInterface $config)
         $this->configs[] = $config;
     }
 
+    /**
+     * @return ConfigInterface[]
```
Comment (stof) [inline Symfony/CS/Fixer.php]: Improving phpdoc is unrelated to adding integration tests. You should submit it as a separate PR

File: Symfony/CS/Tests/IntegrationTest.php [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        $appliedFixers = $this->fixer->fixFileTokens($dummyFile, $tokens, $fixers);
+        if (null === $expected) {
+            if (0 === count($appliedFixers)) {
+                return;
+            }
+            $this->assertSame($input, $tokens->generateCode(), sprintf('No changes expected for "%s" in "%s", fixers applied "%s".', $testTitle, $fileName, implode(', ', $appliedFixers)));
+            // even if the code is the same, the fixers should not have been used
+            $this->fail(sprintf('No changes expected for "%s" in "%s", fixers applied "%s".', $testTitle, $fileName, implode(', ', $appliedFixers)));
+
+            return;
```
Comment (stof) [inline Symfony/CS/Tests/IntegrationTest.php]: no need to return. ``$this->fail()`` already throws an exception

File: Symfony/CS/Tests/IntegrationTest.php [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        $tokens = Tokens::fromCode($input);
+
+        $appliedFixers = $this->fixer->fixFileTokens($dummyFile, $tokens, $fixers);
+        if (null === $expected) {
+            if (0 === count($appliedFixers)) {
+                return;
+            }
+            $this->assertSame($input, $tokens->generateCode(), sprintf('No changes expected for "%s" in "%s", fixers applied "%s".', $testTitle, $fileName, implode(', ', $appliedFixers)));
+            // even if the code is the same, the fixers should not have been used
+            $this->fail(sprintf('No changes expected for "%s" in "%s", fixers applied "%s".', $testTitle, $fileName, implode(', ', $appliedFixers)));
```
Comment (stof) [inline Symfony/CS/Tests/IntegrationTest.php]: why are you always failing for this code path ? This looks weird

File: Symfony/CS/Fixer.php (hunk 3 above)
Comment (stof) [inline Symfony/CS/Fixer.php]: why couldn't the integration test be using the public API of the Fixer to test its whole logic instead of trying to test only a subset of it ?

File: Symfony/CS/Fixer.php (hunk 5 above)
Comment (GrahamCampbell) [inline Symfony/CS/Fixer.php]: :+1:

File: Symfony/CS/Fixer.php (hunk 6 above)
Comment (GrahamCampbell) [inline Symfony/CS/Fixer.php]: Yeh, I ready said that. :)

File: Symfony/CS/Tests/IntegrationTest.php (hunk 7 above)
Comment (GrahamCampbell) [inline Symfony/CS/Tests/IntegrationTest.php]: :+1:

File: Symfony/CS/Tests/IntegrationTest.php (hunk 8 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/IntegrationTest.php]: I'm not sure I understand what you mean here, what would be the ideal path?

Comment (stof) [inline Symfony/CS/Tests/IntegrationTest.php]: well, the point here is that if you have ``null === $expected`` the test *always* fail here

File: Symfony/CS/Fixer.php (hunk 3 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer.php]: Because the fixer reads the file by itself using `file_get_contents` which cannot be stubbed.
I.e. there is no API on the Fixer that supports given code as string and the fixers you want to work on that string.
That is why the AbstractFixerTestBase doesn't use the Fixer itself, but only the implementing FixerInterface object that is being tested.

**Comment by SpacePossum** [comment]
>Why making this a 2.0 feature ? New features can (and should) go in 1.x.

well, I need another API on the Fixer class because this won't work for the `.test` files:
`$new = $old = file_get_contents($file->getRealpath());`
changing the API will be a BC

>I also think you should separate the logic allowing to create integration tests from the actual tests using it in this library, to make the integration testing setup reusable by other packages providing extra fixers. See how Twig does it.

I'll see if I can find it, thanks for the pointers.

File: Symfony/CS/Fixer.php (hunk 3 above)
Comment (stof) [inline Symfony/CS/Fixer.php]: Well, just create a file in a temporary folder for the integration test, which will allow you to run the whole logic.

AbstractFixerTestBase uses only the FixerInterface implementation because it is meant to write unit tests for the FixerInterface implementations, and so should run these classes only, not the whole library

Comment (SpacePossum) [inline Symfony/CS/Fixer.php]: check, done; 7cd692ab8065798f0eaa44958d26dfd79c48fb0f
all changes to `Fixer.php` have been reverted

**Comment by keradus** [comment]
Great idea and I'm very happy to see this PR! ;)

On 2.x line we won't have levels like PSR2 or Symfony - so please put tests into same dir.

**Comment by SpacePossum** [comment]
updated : )

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        $tests = array();
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($fixturesDir), \RecursiveIteratorIterator::LEAVES_ONLY) as $file) {
+            if (!preg_match('/\.test$/', $file)) {
+                continue;
+            }
+
+            $test = file_get_contents($file->getRealpath());
+            $fileName = $file->getFileName();
+            if (preg_match('/--TEST--\s*(.*?)\s*--CONFIG--\s*(.*?)\s*--INPUT--\s?(.*?[\r\n]*)(?:[\s\S]--EXPECT--[\s\S](.*)|$)/s', $test, $match)) {
```
Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: * why `(.*?[\r\n]*)`? What is the difference to just `(.*?)` (when using `s` modifier)?
* `[\s\S]`?? It matches one character which can be a whitespace (`\s`) or not a whitespace (`\S`).

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+                $testTitle = $match[1];
+                $config = $this->getFixersFromConfig($fileName, $match[2]);
+                $input = $match[3];
+                $expected = 5 === count($match) ? $match[4] : null;
```
Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: `isset($match[4])`

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 9 above)
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: `[\s\S]` is replacement for `.*` with multiline flag on JS, which does not have that flag :D

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+    /**
+     * @param Error[] $errors
+     *
+     * @return string
+     */
+    private function implodeErrors(array $errors)
+    {
+        $errorStr = '';
+        /** @var Error $error*/
```
Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Should not be necessary because of the docblock type hint (`Error[]`).

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 9 above)
Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Ah, didn't know that ;)

Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: At the moment these are accepted files:

```
--TEST----CONFIG----INPUT--
```

```
This is my --TEST--file
with a --CONFIG-- and
an --INPUT--section.
```

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 10 above)
Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: shame we can't use `??` ;)

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 9 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: updated in:
https://github.com/SpacePossum/PHP-CS-Fixer/commit/80785a7c95fb6b2f640e550b215d7c77e0dab710

**Comment by SpacePossum** [comment]
more fixes done

**Comment by keradus** [comment]
> https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1375#issuecomment-137393979

What about that? Are your PR already depends on sobe breaking changes?

**Comment by SpacePossum** [comment]
>What about that? Are your PR already depends on sobe breaking changes?

I'm not sure I follow, since I don't know the meaning of 'sobe'.
As stated in the response on the linked comment of Stof, I've created a PR for 1.11,
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1384

**Comment by keradus** [comment]
typo, shoud be 'some' ;)

**Comment by SpacePossum** [comment]
I see :)
No, the tests do not depend on anything new introduced in the 2.x line (or removed for that matter). However  it uses the `errorManager` to check if everything went well and that part (and the `Error` class) have changed in 2.x.

**Comment by keradus** [comment]
And now integration tests are on master too.
Thank you again @SpacePossum !
